### PR TITLE
WI #85 Add support for server requests and client responses

### LINKED ISSUE #85: Add support for server requests and client responses
Currently requests coming from the server are not supported by LSR. Th goal of this issue is to add support for both server requests and client responses which are used in refactoring scenarios.

4 new cases to handle:
- Recording a server request
- Recording a client response
- Replaying a server request
- Replaying a client response

See
- `RecordingModeController.FromServer`
- `RecordingModeController.FromClient`
- `ReplayModeController.FromServer`
- `ReplayModeController.FromClient`
- `ScriptRobotConnectionController.ReplayMessages`
- `RecordingModeController.RecordScriptMessage`

### PR DESCRIPTION
**Pull request by fm-117**
Fixes #85.

Code is mostly duplicated from existing methods for other message types. This is a breaking change for existing TLSP files as the `Result` message type no longer exists. However it is simply fixed by replacing message category "2" (formerly `Result`) by "1" (`Server`).

### CODE REVIEW

File: Solution/TypeCobol.LanguageServer.Robot.Common/Controller/AbstractReplayRobotConnectionController.cs [hunk 1]
```diff
@@ -261,16 +261,32 @@ protected ResponseResultOrError ReplayMessage(string message)
...
+            {
+                lock (IncomingNotifications)
+                {
+                    IncomingNotifications.Add(message_jsonObject.Item1);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Handle a notification event from the ReplayController
```
Comment (efr15) [inline Solution/TypeCobol.LanguageServer.Robot.Common/Controller/AbstractReplayRobotConnectionController.cs]: Handle a **request** event

File: Solution/TypeCobol.LanguageServer.Robot.Common/Controller/ReplayModeController.cs [hunk 2]
```diff
@@ -308,6 +314,24 @@ public override void FromServer(string message)
                             }
                             consumed = true;
                         }
+                        else if (Utilities.Protocol.IsRequest(jsonObject))
+                        {
+                            this.ResultScript.AddMessage(Script.MessageCategory.Server, message);
+                            RaiseRequestEvent(message, jsonObject);
+                            if (StopAtFirstError)
+                            {
+                                if ((ResultScript.messages.Count - 1) < this.SourceScript.messages.Count)
```
Comment (efr15) [inline Solution/TypeCobol.LanguageServer.Robot.Common/Controller/ReplayModeController.cs]: Is it possible to store `ResultScript.messages.Count - 1` in a local variable to improve readability?

File: Solution/TypeCobol.LanguageServer.Robot.Common/Controller/RecordingModeController.cs [hunk 3]
```diff
@@ -404,21 +429,42 @@ protected virtual bool RecordScriptMessage(Script.MessageCategory category, Util
...
+                    // This is a request from the client: store uri to find it back later when response come
                     string id = Utilities.Protocol.GetRequestId(jsonObject);
-                    RequestIdUriMap[id] = uri;
+                    _clientRequestIdToUri.Add(id, uri);
                 }
-                else if (category == Script.MessageCategory.Result && kind == Utilities.Protocol.Message_Kind.Response)
-                {//Rmove the reference to the request
+                else if (kind == Utilities.Protocol.Message_Kind.Response)
+                {
+                    // This is a response from the client, remove from map
```
Comment (efr15) [inline Solution/TypeCobol.LanguageServer.Robot.Common/Controller/RecordingModeController.cs]: What does this case correspond to? The client sent a request and it responds to it itself?
Shouldn't this be `_serverRequestIdToUri.Remove(id)`?

File: Solution/TypeCobol.LanguageServer.Robot.Common/Controller/RecordingModeController.cs [hunk 4]
```diff
@@ -404,21 +429,42 @@ protected virtual bool RecordScriptMessage(Script.MessageCategory category, Util
...
+            {
+                if (kind == Utilities.Protocol.Message_Kind.Request)
+                {
+                    // This is a request from the server: store uri to find it back later when response is sent
+                    string id = Utilities.Protocol.GetRequestId(jsonObject);
+                    _serverRequestIdToUri.Add(id, uri);
+                }
+                else if (kind == Utilities.Protocol.Message_Kind.Response)
+                {
+                    // This is a response from the server, remove from map
```
Comment (efr15) [inline Solution/TypeCobol.LanguageServer.Robot.Common/Controller/RecordingModeController.cs]: What does this case correspond to? The server sent a request and it responds to it itself?
Shouldn't this be `_clientRequestIdToUri.Remove(id)`?

**Review by efr15 — CHANGES_REQUESTED** [review]

File: Solution/TypeCobol.LanguageServer.Robot.Common/Controller/RecordingModeController.cs (hunk 3 above)
Comment (fm-117) [reply Solution/TypeCobol.LanguageServer.Robot.Common/Controller/RecordingModeController.cs]: This case corresponds to a response sent by the client to the server like after receiving an `ApplyWorkspaceEdit` request, the client will send a response telling whether it applied the changes or not.

As for the uri tracking I've just mirrored the code that existed previously for the client-requests and server-responses. So we now have two dictionaries to manage this tracking, but they both work the same.

File: Solution/TypeCobol.LanguageServer.Robot.Common/Controller/RecordingModeController.cs (hunk 4 above)
Comment (fm-117) [reply Solution/TypeCobol.LanguageServer.Robot.Common/Controller/RecordingModeController.cs]: See my previous remark, here we are dealing with more traditional responses coming from the server after a request initiated by the client.

File: Solution/TypeCobol.LanguageServer.Robot.Common/Controller/ReplayModeController.cs (hunk 2 above)
Comment (fm-117) [reply Solution/TypeCobol.LanguageServer.Robot.Common/Controller/ReplayModeController.cs]: done

File: Solution/TypeCobol.LanguageServer.Robot.Common/Controller/AbstractReplayRobotConnectionController.cs (hunk 1 above)
Comment (fm-117) [reply Solution/TypeCobol.LanguageServer.Robot.Common/Controller/AbstractReplayRobotConnectionController.cs]: done

**Review by efr15 — APPROVED** [review]

File: Solution/TypeCobol.LanguageServer.Robot.Common/Controller/RecordingModeController.cs (hunk 3 above)
Comment (fm-117) [reply Solution/TypeCobol.LanguageServer.Robot.Common/Controller/RecordingModeController.cs]: The code was wrong, fixed.

**Pull request merged**